File: src/MotorDriver/src/runMotor.cpp
#include "/home/bertrandt/BOB/library/MotorDriver/include/MotorDriver.h"

int main() {
	float linearVelocity = 0.0;
	float angularVelocity = 0.0;

    // motor(directionPin, PwmPin, minSpeed, maxSpeed)
    MD::Motor leftMotor(31, 32, -255, 255);
    MD::Motor rightMotor(33, 34, -255, 255);
    
    int leftMotorPwmPin = leftMotor.getPwmPin();
    int rightMotorPwmPin = rightMotor.getPwmPin();

    GPIO::PWM leftMotorPWMPin(leftMotorPwmPin, 1000); // for GPIO::BOARD
    GPIO::PWM rightMotorPWMPin(rightMotorPwmPin, 1000);
    
    int leftSpeed = (linearVelocity + angularVelocity) * 100;
    int rightSpeed = (linearVelocity - angularVelocity) * 100;
    
    leftMotorPWMPin.start(0.0);
    rightMotorPWMPin.start(0.0);
    
    leftMotor.setSpeed(leftSpeed);
    int newLeftSpeed = leftMotor.getSpeed();
    
    rightMotor.setSpeed(rightSpeed);
    int newRightSpeed = rightMotor.getSpeed();
    
    leftMotorPWMPin.ChangeDutyCycle(newLeftSpeed);
    rightMotorPWMPin.ChangeDutyCycle(newRightSpeed);
    
    leftMotor.stop();
    rightMotor.stop();
    
    leftMotorPWMPin.stop();
    rightMotorPWMPin.stop();

    GPIO::cleanup();

    return 0;
}

File: py/runMotor.py
import RPi.GPIO as GPIO
import rospy
from geometry_msgs.msg import Twist

# motor(directionPin, PwmPin, minSpeed, maxSpeed)
leftMotor = Motor(31, 32, 0.0, 100.0)
rightMotor = Motor(35, 33, 0.0, 100.0)

leftMotorPwmPin = leftMotor.pwmPin
rightMotorPwmPin = rightMotor.pwmPin

leftMotorDirection = leftMotor.direction
rightMotorDirection = rightMotor.direction

GPIO.setmode(GPIO.BOARD)
GPIO.setup(32, GPIO.OUT)
GPIO.setup(33, GPIO.OUT)
GPIO.setup(31, GPIO.OUT)
GPIO.setup(35, GPIO.OUT)

leftMotorPWMPin = GPIO.PWM(32, 1000)
rightMotorPWMPin = GPIO.PWM(33, 1000)

leftWheel = leftMotor.speed
rightWheel = rightMotor.speed

def messageCallback(cmd_vel):
    linearVelocityX = cmd_vel.linear.x
    angularVelocityZ = cmd_vel.angular.z

    leftWheel = (linearVelocityX - angularVelocityZ) * 100.0
    rightWheel = (linearVelocityX + angularVelocityZ) * 100.0

    leftMotor.speed = leftWheel
    rightMotor.speed = rightWheel

    newLeftWheel = leftMotor.speed
    newRightWheel = rightMotor.speed

    leftMotorPWMPin.ChangeDutyCycle(newLeftWheel)
    rightMotorPWMPin.ChangeDutyCycle(newRightWheel)

    rospy.loginfo("------------------------------------")
    rospy.loginfo("Linear velocity: %f", linearVelocityX)
    rospy.loginfo("Angular velocity: %f", angularVelocityZ)
    rospy.loginfo("Left wheel speed: %f", leftWheel)
    rospy.loginfo("New left wheelSpeed: %f", newLeftWheel)
    rospy.loginfo("Right wheel speed: %f", rightWheel)
    rospy.loginfo("New right wheelSpeed: %f", newRightWheel)

    if linearVelocityX > 0 and angularVelocityZ < 0:
        rospy.loginfo("Forward right")
    elif linearVelocityX > 0 and angularVelocityZ > 0:
        rospy.loginfo("Forward left")
    elif linearVelocityX < 0 and angularVelocityZ < 0:
        rospy.loginfo("Backwards right")
    elif linearVelocityX < 0 and angularVelocityZ > 0:
        rospy.loginfo("Backwards left")
    elif linearVelocityX == 0 and angularVelocityZ > 0:
        rospy.loginfo("Left")
    elif linearVelocityX > 0 and angularVelocityZ == 0:
        rospy.loginfo("Forward")
    elif linearVelocityX < 0 and angularVelocityZ == 0:
        rospy.loginfo("Backwards")
    elif linearVelocityX == 0 and angularVelocityZ < 0:
        rospy.loginfo("Right")
    else:
        leftMotor.stop()
        rightMotor.stop()
        rospy.loginfo("Not moving")

    rospy.loginfo("------------------------------------")

def main():
    leftMotorPWMPin.start(0.0)
    rightMotorPWMPin.start(0.0)

    rospy.init_node("run_motor")
    rospy.Subscriber("cmd_vel", Twist, messageCallback)
    
    rospy.loginfo("Left pwm pin: %d", leftMotorPwmPin)
    rospy.loginfo("Right pwm pin: %d", rightMotorPwmPin)

    rospy.spin()

if __name__ == "__main__":
    main()
